expr/shaper.go

```diff
@@ -0,0 +1,418 @@
+package expr
+
+import (
+	"fmt"
+
+	"github.com/brimsec/zq/zcode"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+	"github.com/brimsec/zq/zson"
+)
+
+// A ShaperTransform represents one of the different transforms that a
+// shaper can apply.
+type ShaperTransform int
+
+const (
+	Cast = ShaperTransform(iota)
+	Fill
+	Crop
+	Order
+)
+
+type step int
+
+const (
+	copyPrimitive = step(iota) // copy field fromIndex from input record
+	copyContainer
+	castPrimitive // cast field fromIndex from input record
+	null          // write null
+	record        // record into record below us
+)
+
+// A shapeBuilder is a recursive data structure encoding a series of
+// copy/cast steps to be carried out over an input record.
+type shapeBuilder struct {
+	op        step
```

Comment:
I think you should change the type name to `op`.

Reply:
done in 4cd9d91